### PR TITLE
Complete Http2Stream after sending Reset or EndOfStream to server

### LINKED ISSUE #1586: HTTP2: Client and server disagree on active stream count
_From @JamesNK on Saturday, August 24, 2019 2:24:08 AM_

Issue found during HTTP2 stress testing. One HttpClient is sending multiple requests in parallel, then canceling them via RST_STREAM sent to the server. For an unknown reason the server and client disagree on the active stream count, and the sever eventually sends RST_STREAM (REFUSED_STREAM) and GOAWAY (STREAM_CLOSED). This issue eventually leads to https://github.com/aspnet/AspNetCore/issues/13405.

Client and server should agree on the active stream count and the client should wait to send new requests, without error.

Wireshark log: https://www.dropbox.com/s/5ndspth4t0qjyj6/clientreset-error.zip?dl=0

Exception on the client:

```
Unhandled exception. System.Net.Http.HttpRequestException: An error occurred while sending the request.
---> System.IO.IOException: The request was aborted.
---> System.IO.IOException: The response ended prematurely, with at least 9 additional bytes expected.
   at System.Net.Http.Http2Connection.ReadAtLeastAsync(Stream stream, Memory`1 buffer, Int32 minReadBytes)
   at System.Net.Http.Http2Connection.EnsureIncomingBytesAsync(Int32 minReadBytes)
   at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
   at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
   at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Grpc.Net.Client.Internal.GrpcCall`2.SendAsync(HttpClient client, HttpRequestMessage message)
   at Grpc.Net.Client.Internal.GrpcCall`2.GetResponseHeadersAsync()
```

Repo is currently not public. Contact @JamesNK or @Tratcher for the repo source code and instruction.

_Copied from original issue: aspnet/AspNetCore#13406_

### PR DESCRIPTION
**Pull request by alnikola**
It changes the order of Complete and Send Reset/Send EndOfStream operations to prevent creation of a new Http2Stream while the old one has not yet written the final frame to wire.

Fixes #1586

### CODE REVIEW

**Comment by alnikola** [comment]
Test failures are not related to this change. All relevant tests including the stress ones are green.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 1]
```diff
@@ -402,19 +413,20 @@ private void Cancel()
...
+                bool complete = false;
 
                 if (_responseCompletionState == StreamCompletionState.InProgress)
                 {
                     _responseCompletionState = StreamCompletionState.Failed;
                     if (_requestCompletionState != StreamCompletionState.InProgress)
                     {
                         sendReset = true;
-                        Complete();
+                        complete = true;
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:429]: Does it even differ from `sendReset`? Do we need an extra variable?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 2]
```diff
@@ -240,12 +240,11 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
...
                         // This should not cause RST_STREAM to be sent because the request is still marked as in progress.
                         bool sendReset;
-                        (signalWaiter, sendReset) = CancelResponseBody();
+                        (signalWaiter, sendReset, _) = CancelResponseBody();
                         Debug.Assert(!sendReset);
 
                         _requestCompletionState = StreamCompletionState.Failed;
                         Complete();
-
                         SendReset();
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: I overlooked that. Fixed now.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs (hunk 1 above)
Comment (alnikola) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Yes, you are right. I removed the redundant variable.

**Review by ManickaP — APPROVED** [review]

**Review by geoffkizer — APPROVED** [review]

**Comment by geoffkizer** [comment]
I do wonder a little if we should just call Complete inside of SendReset. Seems like we always should do this. But I understand if we don't want to make that change at this point.

Some of the HTTP2 stream teardown/reset/complete logic is more complicated than it needs to be for reasons that no longer apply (specifically, the write queue changes, possibly others). It would be nice to take a pass through this (in the future) and clean this up. Maybe we should file an issue for the future?

**Comment by alnikola** [comment]
@geoffkizer I submitted the issue #55994

**Pull request merged**